src/src/tx_main.cpp

```diff
@@ -1182,48 +1179,55 @@ static void HandleUARTout()
 
 static void HandleUARTin()
 {
-  // Read from the USB serial port
-  if (TxUSB->available())
+  if (firmwareOptions.is_airport)
   {
-    if (firmwareOptions.is_airport)
+    auto size = std::min(apInputBuffer.free(), (uint16_t)TxUSB->available());
+    if (size > 0)
     {
-      auto size = std::min(apInputBuffer.free(), (uint16_t)TxUSB->available());
-      if (size > 0)
-      {
-        uint8_t buf[size];
-        TxUSB->readBytes(buf, size);
+      uint8_t buf[size];
+      TxUSB->readBytes(buf, size);
         apInputBuffer.lock();
         apInputBuffer.pushBytes(buf, size);
         apInputBuffer.unlock();
+    }
+    return;
+  }
+
+  // Read from the USB serial port
+  auto size = std::min(uartInputBuffer.free(), (uint16_t)TxUSB->available());
+  if (size > 0)
+  {
+    uint8_t buf[size];
+    TxUSB->readBytes(buf, size);
+
+    // Lets check if the data is Mav and auto change LinkMode
+    // Start the hwTimer since the user might be operating the module as a standalone unit without a handset.
+    if (connectionState == noCrossfire)
+    {
+      if (isThisAMavPacket(buf, size))
+      {
+        config.SetLinkMode(TX_MAVLINK_MODE);
+        UARTconnected();
       }
     }
+    if (config.GetLinkMode() == TX_MAVLINK_MODE)
+    {
+      uartInputBuffer.lock();
+      uartInputBuffer.pushBytes(buf, size);
+      uartInputBuffer.unlock();
+    }
     else
     {
-      auto size = std::min(uartInputBuffer.free(), (uint16_t)TxUSB->available());
-      if (size > 0)
+      // Process data as CRSF bytes
+      for (uint8_t i = 0; i < size; ++i)
```

Comment:
Feels like crsfParser could use a processBytes() and use that here and SerialCRSF.cpp? why loop everywhere.

Reply:
Fair